Add button that exports pending payments as a PDF

dorianzgraggen opened the pull request:
implements #414

dacin21 commented:
- [ ] Is there a reason you deleted the trailing newline of the .json files (e.g. the files get generated automatically and don't have one)? Usually every file should end with a newline.
- [ ] In `payments_list_service.rb` you added a comment: `# TODO: internatinalizaion`.  You mean "localization" (i.e. translating to French)? Is this something that should be done now, or can we leave it for later / never?
- [ ] In `frontend/src/views/payments/tables/ExpenseSheetsReadyForPaymentTable.tsx`: The whole `handleGeneratePdf` and `Button.onClick` seems a bit convoluted. It gets much cleaner if you use the apiURL function and action tags:
```ts
        <a href={props.mainStore!.apiURL('payments_list.pdf', {locale: props.mainStore.locale})} target={'_blank'}>
          <Button
            button props without onClick
          />
        </a>
```
An even better suggestion: add a function `apiLocalizedURL` to the mainStore that calls `apiURL` with the given params plus the locale param  (take from the mainStore). Then you can just do `href={props.mainStore!.apiURL('payments_list.pdf')}`.
- [ ] In `formatExpenseSheets`, converting to json and back seems very hacky. Can't you use map + spread operators?
```ts
return expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: e.user.bank_iban.match(/.{1,4}/g)!.join(' ');
    },
});
```
e.g. see this post: https://stackoverflow.com/a/43376980
- [ ] In the same part, can you put `.match(/.{1,4}/g)!.join(' ');` into a function or lambda and add a one line comment saying what this does? So we get something like this:
```ts

// TODO: explain formatting in blocks of 4
const formatIban = (iban: string) => iban.match(/.{1,4}/g)!.join(' ');

function formatExpenseSheets(expenseSheets: ExpenseSheetListing[]): ExpenseSheetListing[] {
  return expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: formatIban(e.user.bank_iban);
    },
  });
}
```
or even

```ts

// TODO: explain formatting in blocks of 4
const formatIban = (iban: string) => iban.match(/.{1,4}/g)!.join(' ');

const formatExpenseSheets = (expenseSheets: ExpenseSheetListing[]) => 
 expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: formatIban(e.user.bank_iban),
    },
  });
```
(but I don't know how to properly format the later.)

dacin21 reviewed (changes requested):
- [x] Is there a reason you deleted the trailing newline of the .json files (e.g. the files get generated automatically and don't have one)? Usually every file should end with a newline.
- [x] In `payments_list_service.rb` you added a comment: `# TODO: internatinalizaion`.  You mean "localization" (i.e. translating to French)? Is this something that should be done now, or can we leave it for later / never?
- [x] In `frontend/src/views/payments/tables/ExpenseSheetsReadyForPaymentTable.tsx`: The whole `handleGeneratePdf` and `Button.onClick` seems a bit convoluted. It gets much cleaner if you use the apiURL function and action tags:
```ts
        <a href={props.mainStore!.apiURL('payments_list.pdf', {locale: props.mainStore.locale})} target={'_blank'}>
          <Button
            button props without onClick
          />
        </a>
```
An even better suggestion: add a function `apiLocalizedURL` to the mainStore that calls `apiURL` with the given params plus the locale param  (take from the mainStore). Then you can just do `href={props.mainStore!.apiURL('payments_list.pdf')}`.
- [x] In `formatExpenseSheets`, converting to json and back seems very hacky. Can't you use map + spread operators?
```ts
return expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: e.user.bank_iban.match(/.{1,4}/g)!.join(' ');
    },
});
```
e.g. see this post: https://stackoverflow.com/a/43376980
- [x] In the same part, can you put `.match(/.{1,4}/g)!.join(' ');` into a function or lambda and add a one line comment saying what this does? So we get something like this:
```ts

// TODO: explain formatting in blocks of 4
const formatIban = (iban: string) => iban.match(/.{1,4}/g)!.join(' ');

function formatExpenseSheets(expenseSheets: ExpenseSheetListing[]): ExpenseSheetListing[] {
  return expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: formatIban(e.user.bank_iban);
    },
  });
}
```
or even

```ts

// TODO: explain formatting in blocks of 4
const formatIban = (iban: string) => iban.match(/.{1,4}/g)!.join(' ');

const formatExpenseSheets = (expenseSheets: ExpenseSheetListing[]) => 
 expenseSheets.map(e => {
    ...e,
    user: {
        ...e.user,
        bank_iban: formatIban(e.user.bank_iban),
    },
  });
```
(but I don't know how to properly format the later.)

dorianzgraggen commented:
@dacin21 
> Is there a reason you deleted the trailing newline of the .json files (e.g. the files get generated automatically and don't have one)? Usually every file should end with a newline.

Nope, added them back in. This was probably my editor.
<br/>

> In `payments_list_service.rb` you added a comment: `# TODO: internatinalizaion`.  You mean "localization" (i.e. translating to French)? Is this something that should be done now, or can we leave it for later / never?

This is already done, I just forgot to remove the comment.
<br/>   

> In `frontend/src/views/payments/tables/ExpenseSheetsReadyForPaymentTable.tsx`: The whole `handleGeneratePdf` and `Button.onClick` seems a bit convoluted. It gets much cleaner if you use the apiURL function and action tags:

and 

> An even better suggestion: add a function `apiLocalizedURL` to the mainStore that calls `apiURL` with the given params plus the locale param (take from the mainStore). Then you can just do `href={props.mainStore!.apiURL('payments_list.pdf')}`.

implemented.
<br/>

> 
> In the same part, can you put `.match(/.{1,4}/g)!.join(' ');` into a function or lambda and add a one line comment saying what this does? So we get something like this:

implemented.
